1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,229 @@
+- The set of values may be extended later to include `MatchErrorActionPreference`, which should
+  apply the `$ErrorActionPreference` setting to native commands.
+- A conversion will occur between `$PSNativeCommandErrorAction` and `$ErrorActionPreference`
+  values, where `MatchErrorActionPreference` is converted to the current value of
+  `$ErrorActionPreference`
+
+Valid values for `$ErrorActionPreference`
+
+| Value           | Definition
+----------------  | -------------------
+| Break           | Enter the debugger when an error occurs or when an exception is raised.
+| Continue        | (Default) - Displays the error message and continues executing.
+| Ignore          | Suppresses the error message and continues to execute the command. The Ignore value is intended for per-command use, not for use as saved preference. Ignore isn't a valid value for the $ErrorActionPreference variable.
+| Inquire         | Displays the error message and asks you whether you want to continue.
+| SilentlyContinue| No effect. The error message isn't displayed and execution continues without interruption.
+| Stop            | Displays the error message and stops executing. In addition to the error generated, the Stop value generates an ActionPreferenceStopException object to the error stream. stream
+| Suspend         | Automatically suspends a workflow job to allow for further investigation.
+
+The reported error record should be created with the following details:
+
+| Property        | Definition
+----------------  | -------------------
+| exception:      | `ExitCode`, with the exit code of the failed command.
+| error id:       | `"Program {0} failed with unhandled exit code {1}"`, with the command name and the exit code, from resource string `ProgramFailedToComplete`.
+| error category: | `ErrorCategory.NotSpecified`.
+| object:         | exit code
+
+ This does not provide the actual semantics of bash `set -eo pipefail` as Bourne shell-style
+ integration with existing status handling syntax is not implemented.
+
+`Set -eo pipefail` is a combination of `Set -u` and `Set -o pipefail`:
+
+- `Set -u`, a reference to any variable not previously defined results in an error. Similar to
+  Set-StrictMode in PowerShell.
+- `Set -o pipefail`, by default, pipeline success is determined by the last command executed. `Set
+  -o pipefail` returns an error if any command in the pipeline fails.
+
+ As a result of not including the semantics of `Set -eo pipefail`, PowerShell script logic for
+ handling native command exit codes would need to use either `try`..`catch` or existing exit status
+ handling language constructs, according to the setting of this preference variable.
+
+ One way of overriding `$ErrorActionPreference` for a single native command and handling
+ it's exit status explicitly would be to put this logic into a script block and call it
+ with the invocation operator (`&`).
```

Comment:
This paragraph seems like a hanging addition, but could be turned into good scenario examples, like:

```powershell
try
{
    & { $PSNativeCommandErrorAction = 'Stop'; my_command $arg }
}
catch
{
    # Handle $_ here
}
```

Reply:
@rjmholt, re:

> One thing that isn't discussed here is the actual error output of the native command.

Note the pending suggestion to allow `>` to target _variables_, which would enable collecting the stderr lines with `2> variable:errs`, as suggested by @JamesWTruher: https://github.com/PowerShell/PowerShell/issues/4332
What hasn't been fleshed out is the syntax  to _also_ pass the stderr _through_ (the way that `-ErrorVariable errs` implicitly does) and  (less urgently), how to _add_ to a preexisting target variable's value.

Reply:
@rjmholt, re:

> `& { $PSNativeCommandErrorAction = 'Stop'; my_command $arg }`

Perhaps a simper idiom to recommend is the following:

```powershell
my_command $arg || $(throw)  #  Note the need for $(...)
```

Or, for scripts called from the outside:

```powershell
my_command $arg || $(exit $LASTEXITCODE)  #  Note the need for $(...)
```

Unfortunately, we can't be as concise as POSIX-like shells (`my_command $arg || exit`), both because we're saddled with `$(...)` and because `exit` requires `$LASTEXITCODE` in order to relay the exit code (which would otherwise default to `0`).

Conversely, it would be nice to have a concise way to _ignore_ failure _ad hoc_.
The following does that, but it is somewhat obscure:

```powershell
# Ignore non-zero exit code from my_command
my_command $arg || @()
```

POSIX-like shells offer the `:` built-in for this purpose, which is a quiet no-op that always succeeds.
We could consider offering it too.